File: panorama.py
import os

import numpy

import scipy.misc
import scipy.signal
import scipy.special

BASE_DIRECTORY = "/Users/bill/workbench/youvisit/__old__/panoramadetect/component/"
CACHE = BASE_DIRECTORY + "cache/"

"""============================================================

discretize(function, size)

Convert a function to an array of values

@param function   = The function to discretize
@param size       = The size of the kernel to outpu

============================================================"""
def discretize(size, function, args=[]):
  kernel    = numpy.empty(size)
  half_size = size / 2

  for i in range(size):
    kernel[i] = function(i - half_size, *args)

  return kernel



"""============================================================

              /       x      \
error(x) = erf| ____________ |
              |           __ |
              \ sigma * -/ 2 /

============================================================"""
def error(x, sigma):
  sigma_times_sqrt_two = sigma * numpy.power(2, 0.5)
  x_over_sigma_times_sqrt_two = x / sigma_times_sqrt_two

  return -scipy.special.erf(x_over_sigma_times_sqrt_two)



"""============================================================

                      1             -1 * (x * mu)^2
gaussian(x) = ________________ * e^      ___________
                        ______           
              sigma * -/2 * pi           2 * sigma^2

============================================================"""
def gaussian(x, mu, sigma, rescale=1):
  sigma_times_sqrt_two_pi = sigma * numpy.power(2 * numpy.pi, 0.5)
  A                       = 1 / sigma_times_sqrt_two_pi

  x_minus_mu_sqrd         = numpy.power(x - mu, 2)
  two_times_sigma_sqrd    = 2 * numpy.power(sigma, 2)
  B                       = -1 * x_minus_mu_sqrd / two_times_sigma_sqrd

  return A * numpy.power(numpy.e, B) / rescale



"""============================================================

               (x - mu)^3
trinomial(x) = __________ + beta

                sigma^2

============================================================"""
def trinomial(x, mu, sigma, beta, high, low):
  x_minus_mu_cubed  = numpy.power(x - mu, 3)
  sigma_sqrd      = numpy.power(sigma, 2)

  return max(min(x_minus_mu_cubed / sigma_sqrd + beta, high), low)



"""============================================================

                           (x - mu)^2
negated binomial(x) = -1 * __________ + beta

                             sigma^2

============================================================"""
def binomial(x, mu, sigma, beta):
  x_minus_mu_sqrd = numpy.power(x - mu, 2)
  sigma_sqrd      = numpy.power(sigma, 2)
  A = x_minus_mu_sqrd / sigma_sqrd

  return -1 * A + beta



"""============================================================

examinePanorama

Determine if a give image is a panorama

@param filename = The location of the panorama

============================================================"""
def examinePanorama(filename):
  name, extension = os.path.splitext(os.path.basename(filename))

  cache = {
    "U": CACHE + name + ".U.npy",
    "R": CACHE + name + ".R.npy",
    "L": CACHE + name + ".L.npy"
  }

  # Load from cache if cached
  if os.path.isfile(cache["U"]) and os.path.isfile(cache["R"]) and os.path.isfile(cache["L"]) and False:
    U = numpy.load(cache["U"])
    R = numpy.load(cache["R"])
    L = numpy.load(cache["L"])

    width = U.shape[0]
    height = R.shape[0]
    aspectRatio = width / height
  else:
    """
    Load image from file as grayscale
    """
    Image  = scipy.misc.imread(filename, True)
    Image  /= 255.0 # compress range from (0, 255) to (0, 1)

    Image  = numpy.reshape(numpy.ravel(Image, order="F"), (Image.shape[1], Image.shape[0])) # reorder so (X, Y) instead of (Y, X)
    # width & height shortcuts
    width  = Image.shape[0]
    height = Image.shape[1]
    aspectRatio = width / height

    """
    Calculate edge diffs
    """
    # Crop edges
    edge = {
      "left":   numpy.ravel(Image[0 : 1, :]),
      "right":  numpy.ravel(Image[width - 1 : width, :]),
      "upper":  numpy.ravel(Image[ :, 0 : 1]),
      "lower":  numpy.ravel(Image[ :, height - 1 : height])
    }

    # Array of median value from upper and lower (for diffing)
    median = {
      "upper": numpy.empty(width, dtype="float32"),
      "lower": numpy.empty(width, dtype="float32")
    }

    median["upper"].fill(numpy.median(edge["upper"]))
    median["lower"].fill(numpy.median(edge["lower"]))

    # Calculate diff
    diff = {
      "delta upper":  numpy.absolute(numpy.subtract(edge["upper"], median["upper"])),
      "left - right": numpy.absolute(numpy.subtract(edge["left"], edge["right"])),
      "delta lower":  numpy.absolute(numpy.subtract(edge["lower"], median["lower"]))
    }

    # Transform edge diffs to reflect "panoramaness"
    U = diff["delta upper"]
    R = diff["left - right"]
    L = diff["delta lower"]

    numpy.save(CACHE + name + ".U", U)
    numpy.save(CACHE + name + ".R", R)
    numpy.save(CACHE + name + ".L", L)

  kernel = discretize(12, trinomial, [1, 5, 10, 1, 0])
  R = scipy.signal.convolve(R, kernel, "same")

  minU = numpy.empty(width,   dtype="float32")
  minR = numpy.empty(height,  dtype="float32")
  minL = numpy.empty(width,   dtype="float32")

  threshold = 0.01

  for i, value in enumerate(U):
    if i == 0 or value < U[i - 1] + threshold and value > U[i - 1] - threshold:
      minU[i] = 0.0
    else:
      minU[i] = 100.0


  for i, value in enumerate(R):
    minR[i] = binomial(value, mu=-0.17, sigma=0.1, beta=6)

  for i, value in enumerate(L):
    if i == 0 or value < L[i - 1] + threshold and value > L[i - 1] - threshold:
      minL[i] = 0.0
    else:
      minL[i] = 100.0

  nR = numpy.sum(minR) / float(height)
  nUL = (numpy.sum(minU) / float(width) + numpy.sum(minL) / float(width)) / 2

  # Transform aspect ratio to reflect "panoramaness"
  mRr = trinomial(aspectRatio, 3.39, 6, 0, 5, 0)
  mULr = gaussian(aspectRatio, 2, 0.01, 2)

  # Calculate "panoramaness"
  pC = mRr + nR
  pS = (mULr - nUL + pC) / 2

  return [pS, pC]




"""============================================================
===============================================================



 CODE BELOW APPLIES TO __main__ ONLY



===============================================================
============================================================"""

def printLoader(cur, tot, msg):
  percent = float(cur) / float(tot)

  stdout.write("\r\t" + `int(percent * 100)` + "%\t")
  full = "processing >> \033[94m" + msg
  stdout.write(full.ljust(50))
  stdout.write("\033[0m")
  stdout.write("\033[43m")
  for pos in range(25):
    if pos < 25 * percent:
      stdout.write(" ")
    else:
      stdout.write("\033[0m")
      stdout.write(" ")
  
  stdout.write("\033[0m")
  stdout.write("<< end")
  stdout.flush()

def main():
  graphs_file = open(OUTPUT, "w")
  i = 0
  sample_number = len(samples)

  for sample in samples:
    P = examinePanorama(SAMPLE + sample["file"])

    graphs_file.write(`i` + " " + `P` + "\n")

    i += 1

    printLoader(i, sample_number, sample["file"])

  stdout.write("\n\n\t~~~ finished! ~~~\n\n")
  graphs_file.close()

def test():
  s, c = examinePanorama(SAMPLE + samples[178]["file"])
  print s
  print c
  """
  file_diff = open(BASE_DIRECTORY + "log/" + `int(time())` + ".txt", "w")
  for j, pixel in enumerate(panorama):
    file_diff.write(`j` + " " + `pixel` + "\n")

  file_diff.close()
  """

if __name__ == "__main__":
  from samples  import samples
  from sys      import stdout
  from time     import time

  SAMPLE = BASE_DIRECTORY + "sample/"
  OUTPUT = BASE_DIRECTORY + "log/" + `int(time())` + ".txt"

  test()
  # main()

File: panorama.js
(function ()
{
  function discretize(size, closure, args)
  {
    var size = size|0;
    var args = args || null;

    var kernel = zeroes(size);
    var half_size = (size / 2)|0;

    for (var i = size; i -= 1;) {
      var x = (i - half_size)|0;
      kernel[i] = +closure.apply(null, [x].concat(args));
    }

    return kernel;
  }

  function erf(x)
  {
    var a1 =  0.254829592;
    var a2 = -0.284496736;
    var a3 =  1.421413741;
    var a4 = -1.453152027;
    var a5 =  1.061405429;
    var p  =  0.3275911;

    var sign = 1;
    if (x < 0) {
        sign = -1;
    }
    x = Math.abs(x);

    var t = 1.0 / (1.0 + p * x);
    var y = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * Math.exp(-x * x);

    return sign * y;
  }

  function error(x, sigma)
  {
    var sqrttwo = 1.414213562;

    return erf(x / (sigma * sqrttwo));
  }

  function gaussian(x, mu, sigma, rescale)
  {
    var rescale = (rescale || 1);

    var a = 2.506628274;

    return (1 / (sigma * a)) * Math.exp(Math.E, -1 * Math.exp(x - mu, 2) / 2 * Math.exp(sigma, 2)) / rescale;
  }

  function trinomial(x, mu, sigma, beta, upper, lower)
  {
    var y = Math.exp(x - mu, 3) / Math.exp(sigma, 2) + beta;

    return Math.max(Math.min(y, upper), lower);
  }

  function binomial(x, mu, sigma, beta)
  {
    return -1 * Math.exp(x - mu, 2) / Math.exp(sigma, 2) + beta;
  }

  function crop(grid, x, y, w, h)
  {
    var crop = [];

    var yy = y + h;
    var xx = x + w;

    for (var i = x, j = 0; i < xx; i += 1, j += 1) {
      crop[j] = grid[i].slice(y, yy);
    }

    return grid;
  }

  function diff(a, b)
  {
    if (Object.prototype.toString.call(b) === '[object Array]')
      return a.map(function (x, i)
      {
        return -b[i] + x;
      });
    else
      return a.map(function (x, i)
      {
        return -b + x;
      });
  }

  function median(a)
  {
    return a.reduce(function (a, b)
    {
      return a + b;
    }) / a.length;
  }

  function zeroes(l)
  {
    return Array.apply(null, new Array(l)).map(Number.prototype.valueOf, 0);
  }

  function expand(a)
  {
    return [0].concat(a, [0]);
  }

  function pad(a, l)
  {
    if (l > 0) return pad(expand(a), l - 1);
    else return a;
  }

  function convolve_iter(a, k, o)
  {
    return o.map(function (x, i)
    {
      return k.reduce(function (y, z, j)
      {
        return y + a[i + j]*z;
      }, 0) || x;
    });
  }

  function convolve(a, k, full)
  {
    var output = convolve_iter(pad(a, -1 + k.length), k.reverse(), zeroes(-1 + a.length + k.length));

    if (full) return output;
    else return output.slice(0, a.length);
  }

  function RGBtoY(rgb)
  {
    var gamma = 2.2;
    var rgb = rgb.map(function (x)
    {
      return Math.pow(x/255.0, gamma);
    });

    return 0.2126*rgb[0] + 0.7152*rgb[1] + 0.0722*rgb[2];
  }

  function imload(src, callback)
  {
    var img = new Image;
    img.src = src;
    img.onload = callback;
  }

  function imread(img)
  {
    var width = img.width;
    var height = img.height;

    var canvas = document.createElement('canvas');
    canvas.width = width;
    canvas.height = height;

    var context = canvas.getContext('2d');
    context.drawImage(img, 0, 0);

    var data = context.getImageData(0, 0, width, height).data;

    return zeroes(width).map(function (w, x)
    {
      return zeroes(height).map(function (z, y)
      {
        return RGBtoY([
          data[(y*width + x)*4 + 0],
          data[(y*width + x)*4 + 1],
          data[(y*width + x)*4 + 2]
        ])*255;
      });
    });
  }

  var t = 0.01;

  function threshold(total, x, i, array)
  {
    if (i == 0 || x < array[i - 1] + t && x > array[i - 1] - t)
      return total;
    else
      return total + 100.0;
  }

  function examine(img, width, height)
  {
    var aspectRatio = +(width / height);

    var left_edge   = crop(img, 0, 0, 1, height);
    var right_edge  = crop(img, width - 1, 0, 1, height);
    var upper_edge  = crop(img, 0, 0, width, 1);
    var lower_edge  = crop(img, 0, height - 1, width, 1);

    var upper_median = median(upper_edge);
    var lower_median = median(lower_edge);

    var delta_upper = diff(upper_edge, upper_median);
    var delta_lr    = diff(left_edge, right_edge);
    var delta_lower = diff(lower_edge, lower_median);

    var kernel = discretize(12, trinomial, [1, 5, 10, 1, 0]);
    var convolved_lr = convolve(delta_lr, kernel);

    var rU = delta_upper.reduce(threshold, 0);
    var rR = convolved_lr.reduce(function (total, x)
    {
      return total + binomial(x, -0.17, 0.1, 6);
    });
    var rL = delta_lower.reduce(threshold, 0);

    var nR  = rR / height;
    var nUL = (rU + rL)/(width*2);

    var mRr = trinomial(aspectRatio, 3.39, 6, 0, 5, 0);
    var mULr = gaussian(aspectRatio, 2, 0.01, 2);

    pC = mRr + nR;
    pS = (mULr - nUL + pC) / 2;

    return [pC, pS];
  }

  var src = '//localhost:8000/component/lenna.jpg';

  imload(src, function (e) {
    console.log(examine(imread(this), this.width, this.height));
  });

})();
